ocaml/typing/typecore.ml

```diff
@@ -4054,6 +4068,9 @@ let rec is_nonexpansive exp =
   | Texp_extension_constructor _ ->
     false
   | Texp_exclave e -> is_nonexpansive e
+  | Texp_overwrite (exp1, exp2) ->
+      is_nonexpansive exp1 && is_nonexpansive exp2
```

Comment:
I believe that the current code for the `is_nonexpansive` check should be right:
```
let rec is_nonexpansive exp =
  match exp.exp_desc with
  ...
  | Texp_overwrite (exp1, exp2) -> (* overwrite_ exp1 with exp2, where exp1 is the old allocation and exp2 the new one *)
      is_nonexpansive exp1 && is_nonexpansive exp2
  | Texp_hole _ -> true (* keeping a field from the old allocation *)
```

The reason for the `Texp_overwrite` case is that the underlying mutation of `exp1` can not be observed since we have the only reference to it. In fact, a completely valid model for `Texp_overwrite` would be to ignore `exp1` and just allocate a new cell for `exp2`. The reason for the `Texp_hole` case is that it can always be replaced by a field read from the old allocation, which is non-expansive.

Reply:
OK. Good. I'm convinced. Please summarise this in a comment in the code. Thanks for the explanation.